src/persistence/Elsa.Persistence.MongoDb/MongoDbWorkflowDefinitionRepository.cs

```diff
@@ -0,0 +1,105 @@
+        }
+
+        public async Task<int> CountAsync(VersionOptions? version = null, CancellationToken cancellationToken = default)
+        {
+            return await ((IMongoQueryable<WorkflowDefinition>)_dbClient.WorkflowDefinitions.AsQueryable().WithVersion(version)).CountAsync();
+        }
+
+        public async Task DeleteAsync(WorkflowDefinition workflowDefinition, CancellationToken cancellationToken = default)
+        {
+            var filter = GetFilterWorkflowDefinitionId(workflowDefinition.Id);
+            await _dbClient.WorkflowDefinitions.DeleteOneAsync(filter);
+        }
+
+        public async Task<WorkflowDefinition> GetAsync(string workflowDefinitionId, VersionOptions versionOptions, CancellationToken cancellationToken = default)
+        {
+            return await((IMongoQueryable<WorkflowDefinition>)_dbClient.WorkflowDefinitions
+                .AsQueryable()
+                .Where(x => x.WorkflowDefinitionId == workflowDefinitionId)
+                .WithVersion(versionOptions)).FirstOrDefaultAsync();
+        }
+
+        public async Task<WorkflowDefinition> GetByVersionIdAsync(string workflowDefinitionVersionId, CancellationToken cancellationToken = default)
+        {
+            return await((IMongoQueryable<WorkflowDefinition>)_dbClient.WorkflowDefinitions
+                 .AsQueryable()
+                 .Where(x => x.WorkflowDefinitionVersionId == workflowDefinitionVersionId)).FirstOrDefaultAsync();
+        }
+
+        public WorkflowDefinition Initialize(WorkflowDefinition workflowDefinition)
+        {
+            if (workflowDefinition.Version == 0)
+                workflowDefinition.Version = 1;
+
+            if (string.IsNullOrWhiteSpace(workflowDefinition.WorkflowDefinitionVersionId))
+                workflowDefinition.WorkflowDefinitionVersionId = _idGenerator.Generate();
+
+            return workflowDefinition;
+        }
+
+        public async Task<IEnumerable<WorkflowDefinition>> ListAsync(int? skip = null, int? take = null, VersionOptions? version = null, CancellationToken cancellationToken = default)
+        {
+            var query = _dbClient.WorkflowDefinitions
+               .AsQueryable()
+               .WithVersion(version);
+
+            if (skip != null)
+                query = query.Skip(skip.Value);
+
+            if (take != null)
+                query = query.Take(take.Value);
+
+            return await ((IMongoQueryable<WorkflowDefinition>)query).ToListAsync();
+        }
+
+        public async Task SaveAsync(WorkflowDefinition workflowDefinition, CancellationToken cancellationToken = default)
+        {
+            if (workflowDefinition.Id == 0)
+            {
+                // If there is no instance yet, max throws an error
+                if (await _dbClient.WorkflowInstances.AsQueryable().AnyAsync())
+                {
+                    workflowDefinition.Id = await _dbClient.WorkflowDefinitions.AsQueryable().MaxAsync(x => x.Id) + 1;
```

Comment:
So, the reason WorkflowDefinition has a numerical `Id` property is because this is required by YesSQL. But now that you abstracted this way, perhaps we should consider removing the `Id` property as well, and instead have another set of models specific to the YesSQL provider (e.g. WorkflowDefinitionDocument) and use AutoMapper between the domain model and the persistence model?

Reply:
The ID thing has been bothering me as well. I want to avoid AutoMapper if possible, because it costs performance. I think I have an alternative.

Reply:
Agreed, let's do that in a separate PR. Let me know if you're ready for me to merge this one.

Reply:
From my side, I am done with these changes.